src/armature/GuidingVectors.ts

```diff
@@ -0,0 +1,157 @@
+import { coord } from '../calder';
+import { vec3From4 } from '../math/utils';
+import { Mapper } from '../utils/mapper';
+import { Cost, GeneratorInstance } from './Generator';
+import { Node } from './Node';
+
+import 'bezier-js';
+import { vec3, vec4 } from 'gl-matrix';
+import { minBy, range } from 'lodash';
+
+type Closest = {
+    curve: BezierJs.Bezier;
+    point: BezierJs.Projection;
+};
+
+/**
+ * A cost function that doesn't look at the geometry of a model, only the shape, by comparing
+ * the angles in the skeleton to guiding vectors.
+ */
+export class GuidingVectors {
+    private vectors: BezierJs.Bezier[];
+    private nodeLocations: Map<Node, vec4> = new Map<Node, vec4>();
+
+    /**
+     * @param {BezierJs.Bezier[]} guidingVectors The Bezier paths that will be used to guide the
+     * growth of the procedural shape.
+     */
+    constructor(guidingVectors: BezierJs.Bezier[]) {
+        this.vectors = guidingVectors;
+    }
+
+    /**
+     * For debugging/visualization purposes, generates a buffer that is used to render lines in
+     * the vector field.
+     *
+     * @param {number} radius To what distance from the origin field lines should be generated for
+     * @param {number} step The space between field lines
+     * @returns {Float32Array} A buffer of vertices for the vector field lines.
+     */
+    public generateVectorField(radius: number = 3, step: number = 0.5): Float32Array {
```

Comment:
This breaks the Dependency Inversion Principle if this class is supposed to implement the same `CostFn` interface. Do you think we could refactor this somehow?

Reply:
I'm not sure what the best way to do this would be. From SOSMC's point of view, it only uses the `getCost` method, so it only has access to the `CostFn` interface. However, there will inevitably be different things that we want to visualize for each type of cost function so I'm not sure how to get them into the same interface for absolutely everything, so exposing extra methods that SOSMC can't see but that we can still pass to the renderer seemed like a decent compromise to me. I'm open to better suggestions though